igvc_platform/src/imu/main.cpp

```diff
@@ -115,128 +132,136 @@ int main(int argc, char** argv)
+    double yaw_mag = atan2(magnetometer_msg.y, magnetometer_msg.x);
+    ROS_INFO("Magnetometer heading: %f", yaw_mag);
+
+    if (fabs(yaw - yaw_mag) > heading_dev_thresh_deg * M_PI / 180.0) {
+      ROS_WARN("Magnetometer heading and filtered yaw measurement disagree by > %f degrees.", heading_dev_thresh_deg);
+    }
+
+    // Publish sensor messages with corrected transformation.
     sensor_msgs::Imu msg;
     msg.header.frame_id = "imu";
     msg.header.stamp = ros::Time::now();
     msg.header.seq = seq++;
 
-    msg.linear_acceleration.x = rotated_accel[0];
-    msg.linear_acceleration.y = rotated_accel[1];
-    msg.linear_acceleration.z = rotated_accel[2];
-    msg.linear_acceleration_covariance = { 0.01, 1e-6, 1e-6, 1e-6, 0.01, 1e-6, 1e-6, 1e-6, 0.01 };
+    // TODO: Tune values, parameterize.
+    msg.linear_acceleration.x = accel_vec[0];
+    msg.linear_acceleration.y = accel_vec[1];
+    msg.linear_acceleration.z = accel_vec[2];
+    msg.linear_acceleration_covariance = { 0.01, 1e-6, 1e-6,
+                                           1e-6, 0.01, 1e-6,
+                                           1e-6, 1e-6, 0.01 };
 
-    msg.angular_velocity.x = rotated_angular_vel[0];
-    msg.angular_velocity.y = rotated_angular_vel[1];
-    msg.angular_velocity.z = rotated_angular_vel[2];
-    msg.angular_velocity_covariance = { 0.02, 1e-6, 1e-6, 1e-6, 0.02, 1e-6, 1e-6, 1e-6, 0.02 };
+    msg.angular_velocity.x = angular_vel_vec[0];
+    msg.angular_velocity.y = angular_vel_vec[1];
+    msg.angular_velocity.z = angular_vel_vec[2];
+    msg.angular_velocity_covariance = { 0.02, 1e-6, 1e-6,
+                                        1e-6, 0.02, 1e-6,
+                                        1e-6, 1e-6, 0.02 };
 
-    tf::Quaternion pre_rotated(orientation_msg.x, orientation_msg.y, orientation_msg.z, orientation_msg.w);
 
-    double roll, pitch, yaw;
-    tf::Matrix3x3(pre_rotated).getRPY(roll, pitch, yaw);
-    ROS_INFO_STREAM("prerotated " << roll << " " << pitch << " " << yaw);
-    tf::Quaternion rotated = tf::createQuaternionFromRPY(3.14159 / 2, 0, 0) * pre_rotated;
-    rotated = rotated.normalize();
-
-    tf::Matrix3x3(rotated).getRPY(roll, pitch, yaw);
-    ROS_INFO_STREAM("rotated = " << roll << " " << pitch << " " << yaw);
-    tf::Matrix3x3(pre_rotated).getRPY(roll, pitch, yaw);
-    yaw = -yaw;
     geometry_msgs::Quaternion orientation;
-    orientation.x = rotated.x();
-    orientation.y = rotated.y();
-    orientation.z = rotated.z();
-    orientation.w = rotated.w();
-
+    //orientation.x = quaternion_raw.x();
+    //orientation.y = quaternion_raw.y();
+    //orientation.z = quaternion_raw.z();
+    //orientation.w = quaternion_raw.w();
+    // This assumes flat ground and uses magnetometer for absolute heading.
+    // TODO: This assumption may not hold when we switch to 3D lidar, and stop using flat ground assumption
+    tf::Quaternion quaternion_mag;
+    quaternion_mag.setRPY(0,0,yaw_mag);
```

Comment:
We should stop doing this.

Reply:
Will address when we start switching code to handle 3D cases